Description: limit keyname to prevent escape from keydir
 Limit key names to keys within the keydir by refusing keynames containing
 a slash.
Author: Andy Whitcroft <apw@canonical.com>
Forwarded: no
Last-Update: 2019-05-31

---

--- a/lib/rsa/rsa-sign.c
+++ b/lib/rsa/rsa-sign.c
@@ -63,6 +63,11 @@
 	FILE *f;
 	int ret;
 
+	if (strchr(name, '/')) {
+		fprintf(stderr, "Invalid key name '%s': contains '/' \n", name);
+		return -EACCES;
+	}
+
 	*rsap = NULL;
 	snprintf(path, sizeof(path), "%s/%s.crt", keydir, name);
 	f = fopen(path, "r");
@@ -210,6 +215,11 @@
 	RSA *rsa;
 	FILE *f;
 
+	if (strchr(name, '/')) {
+		fprintf(stderr, "Invalid key name '%s': contains '/' \n", name);
+		return -EACCES;
+	}
+
 	*rsap = NULL;
 	snprintf(path, sizeof(path), "%s/%s.key", keydir, name);
 	f = fopen(path, "r");
--- a/tools/kwbimage.c
+++ b/tools/kwbimage.c
@@ -395,6 +395,11 @@
 	if (!keydir)
 		keydir = ".";
 
+	if (strchr(name, '/')) {
+		fprintf(stderr, "Invalid key name '%s': contains '/' \n", name);
+		return -EACCES;
+	}
+
 	snprintf(path, sizeof(path), "%s/%s.key", keydir, name);
 	f = fopen(path, "r");
 	if (!f) {
